### PR TITLE
Support prometheus custom configuration file

### PR DESCRIPTION
**Pull request by yuchaoran2011**
1. Added an API to pass custom Prometheus configuration file contained in a job image. The current APIs only support passing the contents of a custom Prometheus configuration file, but not the file name. The new API allows the user to bake his config file in the Spark job image and reference it in the job YAML.

2. Automatically truncate the volume name to fit 63 chars. This [PR](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/405) introduced a band-aid solution to shorten the suffixes appended by the operator. Now the operator simply truncates the volume name to be 63 characters if it's longer than 63. The user will no longer to experience a failed job because of this issue.

Both of these issues are what we faced in our internal applications using the operator at Lightbend. We think these improvements will enhance the user experience and improve robustness of the operator. 

3. Changed the separator between the namespace and the rest of the metric name from hyphen to dot. Hyphen is really a poor choice of a separator because both the namespace and the metric name can contain hyphens, causing mismatches in the regex. Dot would solve this problem as K8s resource names are not allowed to contain dots.

### CODE REVIEW

File: pkg/config/constants.go [hunk 1]
```diff
@@ -177,84 +177,84 @@ const DefaultPrometheusConfiguration = `
...
     type: GAUGE
     labels:
       app_namespace: "$1"
       app_name: "$2"
-  - pattern: metrics<name=(\S+)-(\S+)\.driver\.(\S+)\.StreamingMetrics\.streaming\.(\S+)><>Value
+  - pattern: metrics<name=(\S+).(\S+)\.driver\.(\S+)\.StreamingMetrics\.streaming\.(\S+)><>Value
     name: spark_streaming_driver_$4
     type: GAUGE
     labels:
       app_namespace: "$1"
```
Comment (skonto) [inline pkg/config/constants.go]: The app_namespace prefix needs to be removed. Spark core metrics does not have this.
The pattern is like this: https://github.com/prometheus/jmx_exporter/blob/master/example_configs/spark.yml#L14

Comment (yuchaoran2011) [reply pkg/config/constants.go]: @liyinan926 Are you ok with me removing the namespace prefix from metric names altogether?

Comment (liyinan926) [reply pkg/config/constants.go]: I prefer keeping them. Even though Spark core metrics in the linked example don't have namespace, we can always support them by adding the namespace as part of the prefix. It's useful to filter metrics by namespaces.

Comment (yuchaoran2011) [reply pkg/config/constants.go]: @liyinan926 Ok. Could you review this PR and get it merged? Thanks!

File: pkg/webhook/webhook.go [hunk 2]
```diff
@@ -256,7 +256,7 @@ func mutatePods(review *admissionv1beta1.AdmissionReview, sparkJobNs string) *ad
 	response := &admissionv1beta1.AdmissionResponse{Allowed: true}
 
 	if !isSparkPod(pod) || !inSparkJobNamespace(review.Request.Namespace, sparkJobNs) {
-		glog.V(2).Info(pod.Name, " in namespace ", review.Request.Namespace, " not mutated")
```
Comment (yuchaoran2011) [inline pkg/webhook/webhook.go]: Somehow `pod.Name` is always empty. But `pod.GetObjectMeta().GetName()` works.

File: pkg/apis/sparkoperator.k8s.io/v1beta1/types.go [hunk 3]
```diff
@@ -478,8 +478,12 @@ type PrometheusSpec struct {
 	// Optional.
 	// If not specified, 8090 will be used as the default.
 	Port *int32 `json:"port"`
+	// ConfigFile is the path to the custom Prometheus configuration file provided in the Spark job image.
+	// ConfigFile takes precedence over Configuration, which is shown below.
+	ConfigFile string `json:"configFile"`
```
Comment (liyinan926) [inline pkg/apis/sparkoperator.k8s.io/v1beta1/types.go]: This should have type `*string` to signal it's optional. We also need to add tag `omitempty` for this and the field below.

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 4]
```diff
@@ -39,50 +40,67 @@ const (
...
+			_, updateErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Update(cm)
+			return updateErr
+		})
 
-	if retryErr != nil {
-		return fmt.Errorf("failed to apply %s in namespace %s: %v", prometheusConfigMapName, app.Namespace, retryErr)
+		if retryErr != nil {
+			return fmt.Errorf("failed to apply %s in namespace %s: %v", prometheusConfigMapName, app.Namespace, retryErr)
+		}
+		javaOption = fmt.Sprintf("-javaagent:%s=%d:%s/%s", app.Spec.Monitoring.Prometheus.JmxExporterJar,
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config.go]: Please add an empty line before this line.

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 5]
```diff
@@ -39,50 +40,67 @@ const (
...
+	}
 
-		cm.Data = configMap.Data
-		_, updateErr := kubeClient.CoreV1().ConfigMaps(app.Namespace).Update(cm)
-		return updateErr
-	})
+	configFile := app.Spec.Monitoring.Prometheus.ConfigFile
+	var javaOption string
+	if configFile != "" {
+		glog.V(2).Infof("Overriding the default Prometheus configuration with config file %s in the Spark job image.", configFile)
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config.go]: In `in the Spark job image`, I think we can remove job here.

File: pkg/webhook/webhook.go (hunk 2 above)
Comment (liyinan926) [reply pkg/webhook/webhook.go]: Interesting.

File: pkg/webhook/patch.go [hunk 6]
```diff
@@ -206,6 +208,11 @@ func addGeneralConfigMaps(pod *corev1.Pod) []*patchOperation {
 	namesToMountPaths := config.FindGeneralConfigMaps(pod.Annotations)
 	for name, mountPath := range namesToMountPaths {
 		volumeName := name + "-vol"
+		if len(volumeName) > maxNameLength {
+			glog.V(2).Infof("ConfigMap volume name %s is too long. Truncating to length %d.", volumeName, maxNameLength)
```
Comment (liyinan926) [inline pkg/webhook/patch.go]: We can combine this and the log message below into one that's after the truncation and indicates the truncation and the result name.

File: pkg/controller/sparkapplication/monitoring_config.go (hunk 4 above)
Comment (yuchaoran2011) [reply pkg/controller/sparkapplication/monitoring_config.go]: Line 77 is restored.

File: pkg/config/constants.go (hunk 1 above)
Comment (skonto) [reply pkg/config/constants.go]: @liyinan926 it does not make sense to add it. Spark core emits these metrics. How would you modify them? This right now creates problems with parsing.

Comment (liyinan926) [reply pkg/config/constants.go]: You can specify prefix for metrics using `spark.metrics.namespace`.

Comment (skonto) [reply pkg/config/constants.go]: Yes but the default shouldnt be that IMHO.

**Pull request merged**